File: sorting_algorithms_code/merge_sort.cpp
// Sort an array in non-decreasing order using merger_sort
#include <iostream>
#include <cstdio>

void merge(int arr[], int left, int mid, int right) {
    // The length of left and right segment
    int l = mid - left + 1;
    int r = right - mid;

    // Create two array for left and right
    int L[l], R[r];
    
    // Fill the L and R
    for(int i = 0; i < l; i++) {
        L[i] = arr[left + i];
    }
    for(int i = 0; i < r; i++) {
        R[i] = arr[mid + i + 1];
    }

    // Merge to the master array
    int i = 0, j = 0, k = left;
    while(i < l && j < r) {
        if(L[i] <= R[j]) {
            arr[k] = L[i];
            i++;
        } else {
            arr[k] = R[j];
            j++;
        }
        k++;
    }
    
    // If L or R is not exhaugsted
    while(i < l) {arr[k++] = L[i++];}
    while(j < r) {arr[k++] = R[j++];}
}

void mergeSort(int arr[], int left, int right) {
    if(left >= right) return;
    int mid = left + (right - left)/2;
    mergeSort(arr, left, mid);
    mergeSort(arr, mid + 1, right);
    merge(arr, left, mid, right);
}

void printArr(int arr[], int size) {
    for (int i = 0; i < size; i++)
        printf("%d ", arr[i]);
    printf("\n");
}

int main() {
    int arr[] = {4,2,3,6,1,8,7};
    int left = 0, right = sizeof(arr)/sizeof(arr[0]) - 1;
    // mergeSort(arr, left, right);
    
    printArr(arr, right + 1);
}

File: 2_sortings_and_bs/merge_sort.py
def merge(arr, l, m, r):
    l_len = m - l + 1
    r_len = r - m

    # Create L and R array
    L = arr[l:m+1]      #[l,m]
    R = arr[m+1:r+1]    #[m+1,r]
    i = j = 0
    k = l
    # Modify arr in-place
    while i < l_len and j < r_len:
        if(L[i] <= R[j]):
            arr[k] = L[i]
            i += 1
        else:
            arr[k] = R[j]
            j += 1
        k += 1

    while i < l_len:
        arr[k] = L[i]
        i += 1
        k += 1
    while j < r_len:
        arr[k] = R[j]
        j += 1
        k += 1

def merge_sort(arr, l, r):
    if(l < r):
        mid = l + (r - l) // 2
        merge_sort(arr, l, mid)
        merge_sort(arr, mid + 1, r)
        merge(arr, l, mid, r)

        